### PR TITLE
Manuall merge of ros2-prealpha to ros2-dev

### PR DESCRIPTION
**Pull request by KmakD**
Manual merge of prealpha branch. Changes regarding `ros2_control` and `panther_hardware_interfaces` were omitted.

### CODE REVIEW

File: panther_gpiod/include/panther_gpiod/gpio_driver.hpp [hunk 1]
```diff
@@ -245,9 +254,10 @@ class GPIODriver
     {GPIOPin::GPIN2, "GPIN2"},
     {GPIOPin::LED_SBC_SEL, "LED_SBC_SEL"},
     {GPIOPin::SHDN_INIT, "SHDN_INIT"},
+    {GPIOPin::STAGE2_INPUT, "STAGE2_INPUT"},
     {GPIOPin::VDIG_OFF, "VDIG_OFF"},
     {GPIOPin::VMOT_ON, "VMOT_ON"},
-  };
+    {GPIOPin::MOTOR_ON, "MOTOR_ON"}};
```
Comment (macstepien) [inline panther_gpiod/include/panther_gpiod/gpio_driver.hpp]: is it formatted correctly? shouldn't `};` be in the next line?

File: panther_gpiod/src/gpio_driver.cpp [hunk 2]
```diff
@@ -105,9 +105,18 @@ gpiod::line_settings GPIODriver::GenerateLineSettings(const GPIOInfo & gpio_info
 {
   auto settings = gpiod::line_settings();
   settings.set_direction(gpio_info.direction);
-  settings.set_output_value(gpio_info.init_value);
   settings.set_active_low(gpio_info.active_low);
 
+  // Set the initial value only when the line is configured for the first time;
+  // otherwise, set the last known value
+  gpiod::line::value new_outpu_value;
```
Comment (macstepien) [inline panther_gpiod/src/gpio_driver.cpp]:
```suggestion
  gpiod::line::value new_output_value;
```

File: panther_gpiod/src/gpio_driver.cpp [hunk 3]
```diff
@@ -217,7 +236,7 @@ void GPIODriver::MonitorAsyncEvents()
   pollfd.events = POLLIN;
 
   {
-    std::lock_guard<std::mutex> lck(monitor_init_mtx_);
+    std::unique_lock<std::mutex> lck(monitor_init_mtx_);
```
Comment (macstepien) [inline panther_gpiod/src/gpio_driver.cpp]: I think that lock_guard is better in this case.

File: panther_gpiod/src/gpio_driver.cpp [hunk 4]
```diff
@@ -192,11 +211,11 @@ void GPIODriver::GPIOMonitorOn()
     }
   }
 
+  std::unique_lock<std::mutex> lck(monitor_init_mtx_);
```
Comment (macstepien) [inline panther_gpiod/src/gpio_driver.cpp:207]: also here, and above (I can't add comment there), lock_guard will be enough

**Review by macstepien — CHANGES_REQUESTED** [review]

File: panther_gpiod/include/panther_gpiod/gpio_driver.hpp (hunk 1 above)
Comment (KmakD) [reply panther_gpiod/include/panther_gpiod/gpio_driver.hpp]: Depends in there is `,` after last element. I copied it as it was in prealpha but it make sense to change that

File: panther_gpiod/src/gpio_driver.cpp (hunk 3 above)
Comment (KmakD) [reply panther_gpiod/src/gpio_driver.cpp]: I don't know the reason this was changed for prealpha. @pidpawel ?

Comment (pkowalsk1) [reply panther_gpiod/src/gpio_driver.cpp]: something went no yes with wild merging. In this simple case lock_guard should be a better fit.

File: panther_bringup/launch/bringup.launch.py [hunk 5]
```diff
@@ -14,20 +14,69 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import textwrap
+
+import click
```
Comment (pkowalsk1) [inline panther_bringup/launch/bringup.launch.py:19]: is it pre-commit?

File: panther_gpiod/src/gpio_driver.cpp [hunk 6]
```diff
@@ -105,9 +105,18 @@ gpiod::line_settings GPIODriver::GenerateLineSettings(const GPIOInfo & gpio_info
...
 
+  // Set the initial value only when the line is configured for the first time;
+  // otherwise, set the last known value
+  gpiod::line::value new_output_value;
+  if (!line_request_) {
+    new_output_value = gpio_info.init_value;
+  } else {
+    new_output_value = gpio_info.value;
+  }
+  settings.set_output_value(new_output_value);
```
Comment (pkowalsk1) [inline panther_gpiod/src/gpio_driver.cpp]:
```suggestion
  gpiod::line::value new_output_value = line_request_ ? gpio_info.value : gpio_info.init_value;
  settings.set_output_value(new_output_value);
```
just a suggestion

File: panther_gpiod/src/gpio_driver.cpp [hunk 7]
```diff
@@ -138,6 +147,16 @@ void GPIODriver::ChangePinDirection(const GPIOPin pin, const gpiod::line::direct
...
+bool GPIODriver::IsPinAvaible(const GPIOPin pin) const
+{
+  for (auto & info : gpio_info_storage_) {
+    if (info.pin == pin) {
+      return true;
+    }
+  }
+  return false;
+}
+
```
Comment (pkowalsk1) [inline panther_gpiod/src/gpio_driver.cpp]:
```suggestion
bool GPIODriver::IsPinAvaible(const GPIOPin pin) const
{
  return std::any_of(gpio_info_storage_.begin(), gpio_info_storage_.end(),
                     [&](const auto &info) { return info.pin == pin; });
}

```
It is considered more idiomatic in modern C++. Just another suggestion :p 

With such changes, we still need to include <algorithm>.

File: panther_bringup/launch/bringup.launch.py (hunk 5 above)
Comment (KmakD) [reply panther_bringup/launch/bringup.launch.py:19]: yes

File: panther_gpiod/src/gpio_driver.cpp (hunk 3 above)
Comment (KmakD) [reply panther_gpiod/src/gpio_driver.cpp]: I think all locks in file can be made `lock_guard` correct me if I'm wrong

**Review by macstepien — APPROVED** [review]

File: panther_gpiod/src/gpio_driver.cpp (hunk 7 above)
Comment (pkowalsk1) [reply panther_gpiod/src/gpio_driver.cpp]: I'm sorry @KmakD. There is a typo in my comment caused by parsing `<>`. With such changes, we still need to include `<algorithm>`.

**Review by pkowalsk1 — APPROVED** [review]

**Pull request merged**